// @buildersgarden/siwa/dist/registry.js
/**
 * registry.ts
 *
 * Agent Identity Registry reader.
 * Provides functions to read agent profiles and reputation from on-chain registries.
 *
 * Dependencies:
 *   npm install viem
 */
import { zeroAddress, createPublicClient, http, encodeFunctionData, parseEventLogs, } from 'viem';
import { getRegistryAddress, getAgentRegistryString, RPC_ENDPOINTS } from './addresses.js';
// ─── ABI Fragments ──────────────────────────────────────────────────
const IDENTITY_REGISTRY_ABI = [
    {
        name: 'ownerOf',
        type: 'function',
        stateMutability: 'view',
        inputs: [{ name: 'tokenId', type: 'uint256' }],
        outputs: [{ name: '', type: 'address' }],
    },
    {
        name: 'tokenURI',
        type: 'function',
        stateMutability: 'view',
        inputs: [{ name: 'tokenId', type: 'uint256' }],
        outputs: [{ name: '', type: 'string' }],
    },
    {
        name: 'getAgentWallet',
        type: 'function',
        stateMutability: 'view',
        inputs: [{ name: 'agentId', type: 'uint256' }],
        outputs: [{ name: '', type: 'address' }],
    },
    {
        name: 'register',
        type: 'function',
        stateMutability: 'nonpayable',
        inputs: [{ name: 'agentURI', type: 'string' }],
        outputs: [{ name: 'agentId', type: 'uint256' }],
    },
    {
        name: 'Registered',
        type: 'event',
        inputs: [
            { name: 'agentId', type: 'uint256', indexed: true },
            { name: 'agentURI', type: 'string', indexed: false },
            { name: 'owner', type: 'address', indexed: true },
        ],
    },
];
const REPUTATION_REGISTRY_ABI = [
    {
        name: 'getSummary',
        type: 'function',
        stateMutability: 'view',
        inputs: [
            { name: 'agentId', type: 'uint256' },
            { name: 'clients', type: 'address[]' },
            { name: 'tag1', type: 'string' },
            { name: 'tag2', type: 'string' },
        ],
        outputs: [
            { name: 'count', type: 'uint64' },
            { name: 'summaryValue', type: 'int128' },
            { name: 'valueDecimals', type: 'uint8' },
        ],
    },
];
// ─── Internal Helpers ───────────────────────────────────────────────
const IPFS_GATEWAY = 'https://gateway.pinata.cloud/ipfs/';
/**
 * Resolve a URI to its JSON content.
 * Supports ipfs://, data:application/json;base64, and http(s):// schemes.
 */
async function resolveURI(uri) {
    if (uri.startsWith('ipfs://')) {
        const cid = uri.slice('ipfs://'.length);
        const response = await fetch(`${IPFS_GATEWAY}${cid}`);
        if (!response.ok)
            throw new Error(`IPFS fetch failed: ${response.status}`);
        return response.json();
    }
    if (uri.startsWith('data:application/json;base64,')) {
        const base64 = uri.slice('data:application/json;base64,'.length);
        const json = Buffer.from(base64, 'base64').toString('utf-8');
        return JSON.parse(json);
    }
    if (uri.startsWith('http://') || uri.startsWith('https://')) {
        const response = await fetch(uri);
        if (!response.ok)
            throw new Error(`HTTP fetch failed: ${response.status}`);
        return response.json();
    }
    throw new Error(`Unsupported URI scheme: ${uri}`);
}
// ─── Public API ─────────────────────────────────────────────────────
/**
 * Read an agent from the Identity Registry and parse its profile.
 *
 * @param agentId  The on-chain agent token ID
 * @param options  Registry address, client, and optional fetchMetadata flag
 */
export async function getAgent(agentId, options) {
    const { registryAddress, client, fetchMetadata = true } = options;
    const [owner, uri, walletAddr] = await Promise.all([
        client.readContract({
            address: registryAddress,
            abi: IDENTITY_REGISTRY_ABI,
            functionName: 'ownerOf',
            args: [BigInt(agentId)],
        }),
        client.readContract({
            address: registryAddress,
            abi: IDENTITY_REGISTRY_ABI,
            functionName: 'tokenURI',
            args: [BigInt(agentId)],
        }),
        client.readContract({
            address: registryAddress,
            abi: IDENTITY_REGISTRY_ABI,
            functionName: 'getAgentWallet',
            args: [BigInt(agentId)],
        }),
    ]);
    const agentWallet = walletAddr === zeroAddress ? null : walletAddr;
    let metadata = null;
    if (fetchMetadata) {
        try {
            const raw = await resolveURI(uri);
            metadata = raw;
        }
        catch {
            metadata = null;
        }
    }
    return { agentId, owner, uri, agentWallet, metadata };
}
/**
 * Read an agent's reputation summary from the Reputation Registry.
 *
 * @param agentId  The on-chain agent token ID
 * @param options  Reputation registry address, client, and optional filters
 */
export async function getReputation(agentId, options) {
    const { reputationRegistryAddress, client, clients = [], tag1 = '', tag2 = '', } = options;
    const result = await client.readContract({
        address: reputationRegistryAddress,
        abi: REPUTATION_REGISTRY_ABI,
        functionName: 'getSummary',
        args: [BigInt(agentId), clients, tag1, tag2],
    });
    const [count, summaryValue, valueDecimals] = result;
    const decimals = Number(valueDecimals);
    const rawValue = BigInt(summaryValue);
    const score = Number(rawValue) / 10 ** decimals;
    return { count: Number(count), score, rawValue, decimals };
}
/**
 * Encode the calldata for an ERC-8004 agent registration without sending it.
 *
 * Use this when your wallet provider handles transaction submission separately
 * (e.g. Bankr's `/agent/submit`, or any ERC-4337 bundler).
 *
 * @example
 * ```typescript
 * import { encodeRegisterAgent } from '@buildersgarden/siwa/registry';
 *
 * const { to, data } = encodeRegisterAgent({
 *   agentURI: 'data:application/json;base64,...',
 *   chainId: 84532,
 * });
 *
 * // Submit via your provider (e.g. Bankr)
 * await fetch('https://api.bankr.bot/agent/submit', {
 *   method: 'POST',
 *   headers: { 'Content-Type': 'application/json', 'X-API-Key': apiKey },
 *   body: JSON.stringify({
 *     transaction: { to, data, value: '0', chainId: 84532 },
 *     waitForConfirmation: true,
 *   }),
 * });
 * ```
 */
export function encodeRegisterAgent(options) {
    const { agentURI, chainId } = options;
    const registryAddress = getRegistryAddress(chainId);
    const data = encodeFunctionData({
        abi: IDENTITY_REGISTRY_ABI,
        functionName: 'register',
        args: [agentURI],
    });
    return { to: registryAddress, data };
}
/**
 * Register an agent on the ERC-8004 Identity Registry in a single call.
 *
 * Builds, signs (via the provided signer), and broadcasts the `register(agentURI)`
 * transaction, then waits for confirmation and parses the `Registered` event.
 *
 * @example
 * ```typescript
 * import { registerAgent, createLocalAccountSigner } from '@buildersgarden/siwa';
 * import { privateKeyToAccount } from 'viem/accounts';
 *
 * const account = privateKeyToAccount('0x...');
 * const signer = createLocalAccountSigner(account);
 *
 * const result = await registerAgent({
 *   agentURI: 'ipfs://...',
 *   chainId: 84532,
 *   signer,
 * });
 * ```
 */
export async function registerAgent(options) {
    const { agentURI, chainId, signer } = options;
    const registryAddress = getRegistryAddress(chainId);
    const rpcUrl = options.rpcUrl || RPC_ENDPOINTS[chainId];
    if (!rpcUrl) {
        throw new Error(`No RPC URL provided and no default endpoint for chain ${chainId}.`);
    }
    const publicClient = createPublicClient({ transport: http(rpcUrl) });
    const address = await signer.getAddress();
    const data = encodeFunctionData({
        abi: IDENTITY_REGISTRY_ABI,
        functionName: 'register',
        args: [agentURI],
    });
    const nonce = await publicClient.getTransactionCount({ address });
    const feeData = await publicClient.estimateFeesPerGas();
    const gasEstimate = await publicClient.estimateGas({
        to: registryAddress,
        data,
        account: address,
    });
    const gas = (gasEstimate * 120n) / 100n;
    const txReq = {
        to: registryAddress,
        data,
        nonce,
        chainId,
        type: 2,
        maxFeePerGas: feeData.maxFeePerGas,
        maxPriorityFeePerGas: feeData.maxPriorityFeePerGas,
        gas,
    };
    const signedTx = await signer.signTransaction(txReq);
    const txHash = await publicClient.sendRawTransaction({
        serializedTransaction: signedTx,
    });
    const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
    const logs = parseEventLogs({
        abi: IDENTITY_REGISTRY_ABI,
        logs: receipt.logs,
        eventName: 'Registered',
    });
    if (logs.length === 0) {
        throw new Error(`Registration tx ${txHash} succeeded but no Registered event was found.`);
    }
    const agentId = logs[0].args.agentId.toString();
    const agentRegistry = getAgentRegistryString(chainId);
    return { agentId, txHash, registryAddress, agentRegistry };
}
